src/encoder/encoderopus.cpp

```diff
@@ -0,0 +1,439 @@
+// See https://tools.ietf.org/html/rfc6716#section-3.2
+const int kMaxOpusBufferSize = 1+1275;
+// Opus frame duration in milliseconds. Fixed to 60ms
+const int kOpusFrameMs = 60;
+
+const mixxx::Logger kLogger("EncoderOpus");
+}
+
+EncoderOpus::EncoderOpus(EncoderCallback* pCallback)
+    : m_bitrate(0),
+      m_bitrateMode(0),
+      m_channels(0),
+      m_samplerate(0),
+	  m_readRequired(0),
+      m_pCallback(pCallback),
+      m_pFifoBuffer(nullptr),
+      m_pFifoChunkBuffer(nullptr),
+      m_pOpus(nullptr),
+      m_header_write(false),
+      m_packetNumber(0),
+      m_granulePos(0) {
+    m_opusComments.insert("ENCODER", "mixxx/libopus");
+    m_pOpusDataBuffer = new unsigned char[kMaxOpusBufferSize]();
+    ogg_stream_init(&m_oggStream, rand());
+}
+
+EncoderOpus::~EncoderOpus() {
+    if (m_pOpus)
+        opus_encoder_destroy(m_pOpus);
+
+    ogg_stream_clear(&m_oggStream);
+    delete[] m_pOpusDataBuffer;
+    delete m_pFifoChunkBuffer;
+    delete m_pFifoBuffer;
+}
+
+void EncoderOpus::setEncoderSettings(const EncoderSettings& settings) {
+    m_bitrate = settings.getQuality();
+    m_bitrateMode = settings.getSelectedOption(EncoderOpusSettings::BITRATE_MODE_GROUP);
+    switch (settings.getChannelMode()) {
+        case EncoderSettings::ChannelMode::MONO:
+            m_channels = 1;
+            break;
+        case EncoderSettings::ChannelMode::STEREO:
+            m_channels = 2;
+            break;
+        case EncoderSettings::ChannelMode::AUTOMATIC:
+            m_channels = 2;
+            break;
+    }
+}
+
+int EncoderOpus::initEncoder(int samplerate, QString errorMessage) {
+    (void)errorMessage;
+    int result;
+
+    if (samplerate != 48000) {
+        kLogger.warning() << "initEncoder failed: samplerate not supported by Opus";
+
+        QString invalidSamplerate = QObject::tr(
+                "Using Opus at samplerates other than 48 kHz "
+                "is not supported by the Opus encoder. Please use "
+                "48000 Hz in \"Sound Hardware\" preferences "
+                "or switch to a different encoding.");
+
+        ErrorDialogProperties* props = ErrorDialogHandler::instance()->newDialogProperties();
+        props->setType(DLG_WARNING);
+        props->setTitle(QObject::tr("Encoder"));
+        props->setText(invalidSamplerate);
+        props->setKey(invalidSamplerate);
+        ErrorDialogHandler::instance()->requestErrorDialog(props);
+        return -1;
+    }
+
+    m_samplerate = samplerate;
+    m_pOpus = opus_encoder_create(m_samplerate, m_channels, OPUS_APPLICATION_AUDIO, &result);
+
+    if (result != OPUS_OK) {
+        kLogger.warning() << "opus_encoder_create failed:" << opusErrorString(result);
+        return -1;
+    }
+
+    // Optimize encoding for high-quality music
+    opus_encoder_ctl(m_pOpus, OPUS_SET_COMPLEXITY(10)); // Highest setting
+    opus_encoder_ctl(m_pOpus, OPUS_SET_SIGNAL(OPUS_SIGNAL_MUSIC));
+
+    if(m_bitrateMode == 0) {
```

Comment:
This should be an enum or at least the constants should explicitly be defined somewhere. As readable code and not only hidden inside comments. Code is validated by the compiler, comments are not.